File: 백준/Gold/11505. 구간 곱 구하기/구간 곱 구하기.py
from sys import stdin
from math import ceil, log2
input = stdin.readline

def init(node,st,ed):
    if st == ed:
        tree[node] = num[st]
        return tree[node]
    mid = (st + ed) // 2
    tree[node] = init(2*node,st,mid) * init(2*node+1,mid+1,ed) % 1000000007
    return tree[node]

def segment_mul(node,st,ed,l,r):
    if l > ed or r < st:
        return 1
    if l <= st and ed <= r:
        return tree[node]
    mid = (st + ed) // 2
    return segment_mul(node*2,st,mid,l,r) * segment_mul(node*2+1,mid+1,ed,l,r) % 1000000007

def update(node,st,ed,idx,diff):
    if idx < st or idx > ed:
        return
    if st == ed:
        tree[node] = diff
    else:
        mid = (st + ed) // 2
        update(node*2,st,mid,idx,diff)
        update(node*2+1,mid+1,ed,idx,diff)
        tree[node] = tree[2*node] * tree[2*node+1] % 1000000007

n,m,k = map(int,input().split())

h = ceil(log2(n))
tree_size = 1 << (h+1)
tree = [0] * tree_size

num = []
for _ in range(n):
    num.append(int(input()))
    
init(1,0,n-1)
    
for _ in range(m+k):
    a,b,c = map(int,input().split())
    if a == 1:
        update(1,0,n-1,b-1,c)
    elif a == 2:
        print(segment_mul(1,0,n-1,b-1,c-1))

File: 백준/Gold/11505. 구간 곱 구하기/구간 곱 구하기.java
import java.io.*;
import java.util.*;

public class Main {
	static int mod = (int) (1e9 + 7);
	static long tree[];

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int K = Integer.parseInt(st.nextToken());
		int h = (int) Math.ceil(Math.log(N) / Math.log(2));
		int treeSize = 1 << (h + 1);
		int startIdx = treeSize / 2 - 1;
		tree = new long[treeSize + 1];
		Arrays.fill(tree, 1);
		for (int i = startIdx + 1; i <= startIdx + N; i++) {
			tree[i] = Long.parseLong(br.readLine());
		}
		init(treeSize);
		for (int i = 0; i < M + K; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			if (a == 1) {
				change(b + startIdx, c);
			} else {
				sb.append(calc(b + startIdx, c + startIdx)).append("\n");
			}
		}
		System.out.print(sb);
	}

	static void init(int i) {
		while (i != 1) {
			tree[i / 2] = tree[i / 2] * tree[i] % mod;
			i--;
		}
	}

	static void change(int idx, int val) {
		tree[idx] = val;
		while (idx > 1) {
			idx /= 2;
			tree[idx] = tree[idx * 2] % mod * tree[idx * 2 + 1] % mod;
		}
	}

	static long calc(int s, int e) {
		long value = 1;
		while (s <= e) {
			if (s % 2 == 1) {
				value = value * tree[s] % mod;
				s++;
			}
			if (e % 2 == 0) {
				value = value * tree[e] % mod;
				e--;
			}
			s = s / 2;
			e = e / 2;
		}
		return value % mod;
	}

}